lsp-mode.el

```diff
@@ -3706,73 +3706,94 @@ and the position respectively."
   "Get lsp completions."
-  (let* ((bounds (bounds-of-thing-at-point 'word-strictly))
-         (trigger-chars (->> (lsp--server-capabilities)
-                             (gethash "completionProvider")
-                             (gethash "triggerCharacters")))
-         start-point result done?)
-    (list
-     (if bounds (car bounds) (point))
-     (point)
-     (lambda (_probe _pred action)
-       (cond
-        ((eq action 'metadata)
-         `(metadata . ((display-sort-function . (lambda (candidates)
-                                                  (--sort (string-lessp (get-text-property 0 'lsp-sort-text it)
-                                                                        (get-text-property 0 'lsp-sort-text other))
-                                                          candidates)))
-                       (category . lsp-substring))))
-        ((eq (car-safe action) 'boundaries) nil)
-        (t (if done?
-               result
-             (setf start-point (point))
-             (let* ((resp (lsp-request-while-no-input "textDocument/completion"
-                                                      (plist-put (lsp--text-document-position-params)
-                                                                 :context (ht ("triggerKind" 1)))))
-                    (items (lsp--sort-completions (cond
-                                                   ((seqp resp) resp)
-                                                   ((hash-table-p resp) (gethash "items" resp))))))
-               (setf done? (or (seqp resp)
-                               (not (gethash "isIncomplete" resp)))
-                     result (seq-into (seq-map (-lambda ((item &as &hash "label"
-                                                               "insertText" insert-text
-                                                               "sortText" sort-text))
-                                                 (propertize (or label insert-text)
-                                                             'lsp-completion-item item
-                                                             'lsp-sort-text sort-text))
-                                               items)
-                                      'list)))))))
-
-     :annotation-function #'lsp--annotate
-     :company-require-match 'never
-     :company-prefix-length (point)
-     :exit-function
-     (lambda (candidate _status)
-       (-let [(&hash "insertText" insert-text
-                     "textEdit" text-edit
-                     "insertTextFormat" insert-text-format
-                     "additionalTextEdits" additional-text-edits)
-              (lsp--resolve-completion (plist-get (text-properties-at 0 candidate) 'lsp-completion-item))]
+  (when (or (--some (lsp--client-completion-in-comments? (lsp--workspace-client it))
+                    (lsp-workspaces))
+            (not (nth 4 (syntax-ppss))))
+    (let* ((trigger-chars (->> (lsp--server-capabilities)
```

Comment:
We are not allowed to do that here, as per docs:

```
NOTE: These functions should be cheap to run since they're sometimes
run from `post-command-hook'; and they should ideally only choose
which kind of completion table to use, and not pre-filter it based
on the current text between START and END (e.g., they should not
obey `completion-styles').
```

We should either by creating a new completion style or by doing that in the `(lambda (probe pred action)` function.

Reply:
Hmm, we need server response to determine the bound, and it's not recommended to have long running code here.
I don't know how we can progress from here.
We need to return `(start end collection . props)` quickly for every call to completion at point, so determine `start` and `end` using server response seems to be out of reach now.

Reply:
It needs the bounds to calculate the prefix to filter against, but since we will create our completion style we could ignore the prefix provided by completion framework and filter against the properly calculated prefix by ourselves which is calculated before the filtering has started. So we could return `(point) (point)` for bounds and then do the right thing when performing filtering.

Reply:
I see, since we've provided completion table as lambda prior to this PR, that function return value will be used as it without intervene of completion style functions. We have to call `completion-do-completions` by ourselves.

With that being said, we still need some reasonable  bounds when `textEdit` is not available (like `rls`)

Reply:
I am not sure I understand 100% your comment:

Prior to this PR we are using this to perform the filtering:

```
(add-to-list 'completion-category-defaults '(lsp-substring (styles . (substring))))
```

We could create our filtering styles, e.g. lsp-fuzzy/lsp-substring/lsp-prefix.

As a side note, if you remove that line from lsp-mode company will display the items without any filtering, so we could do the filtering in the lambda itself and bypass the completion styles. 


> With that being said, we still need some reasonable bounds when textEdit is not available (like rls)

For these we will use `(thing-at-point 'symbol)` but this could be determined after the results are received. Do you mean that you could pass it as bounds by default and then use it only if you don't have textEdit? If yes, this seems to be ok since it leads to the same behaviour.